SECURITY.md: security policy for Solidity (#8390)
ethereum/solidity: SECURITY.md

# Security Policy

## Supported Versions

> Use this section to tell people about which versions of your project are

Author: I believe all Solidity versions are supposed to get security updates.

Reviewer: I don't think so. Only the latest active tree gets updates. In exceptional circumstances, when a breaking release is very young (less then 3-4 releases in), we did issue security updates to the older tree.

Reviewer 2: Yeah, it would be too large of a maintenance burden to support _all_ versions. At most we could have some LTS versions, but that probably only starts making sense once we reached v1.0

Author: Ah I see, I remember some security fixes getting back ported as far back as `0.4.x` but perhaps that was one off?

> Only the latest active tree gets updates

Does this mean the version in which the bug was found and backwards until the last breaking release. So for example, if bug was found in `0.6.2`, then `0.6.0` up to `0.6.2` get patched?

Reviewer: No, just a new release is made in the tree. E.g. since we are early in 0.6.x, it may make sense to make a 0.5.x bugfix release. I think however these cases are very rare if keep our release schedule, because people will transition to new breaking releases more quickly (that's the hope).

The 0.4.x vs 0.5.x was a special case, because 0.4.x was so widely used and it took a really long time to release 0.5.x

currently being supported with security updates.

| Version | Supported          |
| ------- | ------------------ |

> | 5.1.x   | :white_check_mark: |

Reviewer: These do not seem to be correct versions. We only have `0.x.y` so far.

Author: Right, this is simply a copy-paste of the boilerplate text. Will be changed according to the discussion in this thread.

| 5.0.x   | :x:                |
| 4.0.x   | :white_check_mark: |
| < 4.0   | :x:                |

## Reporting a Vulnerability

> Use this section to tell people how to report a vulnerability.

Author: IIUC, here we briefly describe the correct way of reporting vulnerabilities. Is the Ethereum bounty program the correct way of reporting security vulnerabilities (see https://docs.google.com/forms/d/e/1FAIpQLSegLOqYnWv4hWM4RKjDiHIzhZQMavRrKllPJzYJYIPntAim7g/viewform?c=0&w=1)

We probably also need to scope out a vulnerability: This means what is in scope and what qualifies a bug as a security vulnerability?

Reviewer 3: Yes, tell them about the bug bounty programme. In general, I would say that rules of the bug bounty programme have precedence and this here is only to have a more detailed technical discussion.

# Security Policy

The Solidity team and community take all security bugs in Solidity seriously.
We appreciate your efforts and responsible disclosure and will make every effort to acknowledge your contributions.

## Scope

Bugs in the solidity repository are in scope.
Bugs in third-party dependencies e.g., jsoncpp, boost etc. are not in scope unless they result in a solidity specific bug.

Only bugs that have a demonstrable security impact on smart contracts are in scope.

> For example, a solidity program whose optimization is incorrect (e.g., leads to an incorrect output) qualifies as a security bug.

Reviewer 3: ```suggestion
For example, a Solidity program whose optimization is incorrect (e.g., leads to an incorrect output) qualifies as a security bug.
```
(please also check the other mentions of solidity)

Please note that the [rules][2] of the [Ethereum bounty program][1] have precedence over this security policy.

## Supported Versions

As a general rule, only the latest release gets security updates.
Exceptions may be made when the current breaking release is relatively new, e.g. less than three months old.
If you are reporting a bug, please state clearly the Solidity version(s) it affects.

> Example 1: Assuming current release is `0.6.3` and a security bug has been found in it that affects both `0.5.x` and `0.6.x` trees, we may not only patch `0.6.3` (the bug-fix release numbered `0.6.4`) but `0.5.x` as well (the bug-fix release numbered `0.5.(x+1)`).

Reviewer 3: ```suggestion
Example 1: Assuming the current release is `0.6.3` and a security bug has been found in it that affects both `0.5.x` and `0.6.x` trees, we may not only patch `0.6.3` (the bug-fix release numbered `0.6.4`) but `0.5.x` as well (the bug-fix release numbered `0.5.(x+1)`).
```

Example 2: Assuming current release is `0.6.25` and a security bug has been found in it, we may only patch `0.6.25` (in the bug-fix release numbered `0.6.26`) even if the bug affects a previous tree such as `0.5.x`.

> 

Reviewer 3: Should we also mention the bug list?

Author: Yes, that's a good idea. I'll draft a couple of sentences to that effect.

Author: Updated PR.

# Security Policy

The Solidity team and community take all security bugs in Solidity seriously.
We appreciate your efforts and responsible disclosure and will make every effort to acknowledge your contributions.

## Scope

Bugs in the solidity repository are in scope.
Bugs in third-party dependencies e.g., jsoncpp, boost etc. are not in scope unless they result in a solidity specific bug.

> Only bugs that have a demonstrable security impact are in scope.

Reviewer 3: ```suggestion
Only bugs that have a demonstrable security impact on smart contracts are in scope.
```
(security impact on the developer's machine is out of scope)

For example, a solidity program whose optimization is incorrect (e.g., leads to an incorrect output) qualifies as a security bug.
Please note that the [rules][2] of the [Ethereum bounty program][1] have precedence over this security policy.

## Supported Versions

> As a general rule, only the current breaking release gets security updates.

Reviewer 3: ```suggestion
As a general rule, only the latest release gets security updates.
```

> Exceptions may be made when the current breaking release is relatively new i.e., has fewer non-breaking releases since the breaking release.

Reviewer 3: ```suggestion
Exceptions may be made when the current breaking release is relatively new, e.g. less than three months old.
```

If you are reporting a bug, please state clearly the Solidity version(s) it affects.

Example 1: Assuming current release is `0.6.3` and a security bug has been found in it that affects both `0.5.x` and `0.6.x` trees, we may not only patch `0.6.3` (the bug-fix release numbered `0.6.4`) but `0.5.x` as well (the bug-fix release numbered `0.5.(x+1)`).

Example 2: Assuming current release is `0.6.25` and a security bug has been found in it, we may only patch `0.6.25` (in the bug-fix release numbered `0.6.26`) even if the bug affects a previous tree such as `0.5.x`.

## Reporting a Vulnerability

> Report security bugs by emailing `solidity` at `ethereum.org`.

Reviewer 3: Hm, I would say "follow the instructions on https://bounty.ethereum.org/".

## Reporting a Vulnerability

To report a vulnerability, please follow the instructions stated in the [Ethereum bounty program][1].

In the bug report, please include all details necessary to reproduce the vulnerability such as:

- Input program that triggers the bug
- Compiler version affected
- Target EVM version
- Framework/IDE if applicable
- EVM execution environment/client if applicable
- Operating system

Please include steps to reproduce the bug you have found in as much detail as possible.

Once we have received your bug report, we will try to reproduce it and provide a more detailed response.
Once the reported bug has been successfully reproduced, the Solidity team will work on a fix.

The Solidity team maintains the following JSON-formatted lists of patched security vulnerabilities:

- [Summary of known security vulnerabilities][3]
- [List of security vulnerabilities affecting a specific version of the compiler][4].

[1]: https://bounty.ethereum.org/
[2]: https://bounty.ethereum.org/#rules

> [3]: https://solidity.readthedocs.io/en/latest/bugs.html

Reviewer 3: ```suggestion
[3]: https://solidity.readthedocs.io/en/develop/bugs.html
```
